File: deploy/docker/canvasbench/context/cline/cline-wrapper.mjs
#!/usr/bin/env node
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";

const baseURL = (process.env.OPENAI_BASE_URL || "").replace(/\/+$/, "");
const apiKey = process.env.OPENAI_API_KEY || "";
if (!baseURL || !apiKey) throw new Error("canvasbench cline contract requires OPENAI_BASE_URL and OPENAI_API_KEY");
const parsed = new URL(baseURL);
if (!["http:", "https:"].includes(parsed.protocol) || parsed.pathname.replace(/\/+$/, "") !== "/v1") {
  throw new Error("canvasbench cline contract requires a valid OpenAI /v1 base URL");
}

const state = fs.mkdtempSync(path.join(os.tmpdir(), "canvasbench-cline-"));
try {
  const settings = path.join(state, "settings");
  fs.mkdirSync(settings, { mode: 0o700 });
  const provider = {
    provider: "openai-compatible",
    model: process.env.CANVASBENCH_ADAPTER_MODEL || "",
    apiKey,
    baseUrl: baseURL,
    protocol: "openai-chat",
    client: "openai-compatible",
    contextWindow: Number(process.env.CANVASBENCH_CONTEXT_WINDOW || 32768),
    maxTokens: Number(process.env.CANVASBENCH_MAX_OUTPUT_TOKENS || 4096),
    capabilities: ["streaming", "tools"],
  };
  const providers = {
    version: 1,
    lastUsedProvider: "openai-compatible",
    providers: {
      "openai-compatible": {
        settings: provider,
        updatedAt: "2026-01-01T00:00:00.000Z",
        tokenSource: "manual",
      },
    },
  };
  const providerPath = path.join(settings, "providers.json");
  fs.writeFileSync(providerPath, `${JSON.stringify(providers)}\n`, { mode: 0o600 });
  const args = process.argv.slice(2);
  const completed = spawnSync("/opt/harness/cline-real", args, {
    env: { ...process.env, CLINE_DATA_DIR: state, CLINE_PROVIDER_SETTINGS_PATH: providerPath },
    encoding: "utf8",
    maxBuffer: 32 * 1024 * 1024,
  });
  if (completed.error) throw completed.error;
  const exitCode = completed.status ?? 1;
  const captured = `${completed.stdout || ""}${completed.stderr || ""}`.trim();
  const output = captured || `cline exited ${exitCode} without output`;
  const session = crypto.createHash("sha256").update(args.join("\0")).digest("hex").slice(0, 24);
  process.stdout.write(`${JSON.stringify({
    schema: "canvasbench.cli-wrapper.v1",
    harness: "cline",
    session_id: `cline-${session}`,
    output,
    exit_code: exitCode,
  })}\n`);
  process.exitCode = exitCode;
} finally {
  fs.rmSync(state, { recursive: true, force: true });
}

File: deploy/docker/canvasbench/context/cline/cline-wrapper.py
#!/usr/bin/env python3
import hashlib
import json
import os
from pathlib import Path
import shutil
import subprocess
import sys
import tempfile
from urllib.parse import urlsplit


base_url = os.environ.get("OPENAI_BASE_URL", "").rstrip("/")
api_key = os.environ.get("OPENAI_API_KEY", "")
if not base_url or not api_key:
    raise SystemExit("canvasbench cline contract requires OPENAI_BASE_URL and OPENAI_API_KEY")
parsed = urlsplit(base_url)
if parsed.scheme not in {"http", "https"} or not parsed.netloc or parsed.path.rstrip("/") != "/v1":
    raise SystemExit("canvasbench cline contract requires a valid OpenAI /v1 base URL")

state = Path(tempfile.mkdtemp(prefix="canvasbench-cline-", dir="/tmp"))
try:
    settings = state / "settings"
    settings.mkdir(mode=0o700)
    provider = {
        "provider": "openai-compatible",
        "model": os.environ.get("CANVASBENCH_ADAPTER_MODEL", ""),
        "apiKey": api_key,
        "baseUrl": base_url,
        "protocol": "openai-chat",
        "client": "openai-compatible",
        "contextWindow": int(os.environ.get("CANVASBENCH_CONTEXT_WINDOW", "32768")),
        "maxTokens": int(os.environ.get("CANVASBENCH_MAX_OUTPUT_TOKENS", "4096")),
        "capabilities": ["streaming", "tools"],
    }
    providers = {
        "version": 1,
        "lastUsedProvider": "openai-compatible",
        "providers": {
            "openai-compatible": {
                "settings": provider,
                "updatedAt": "2026-01-01T00:00:00.000Z",
                "tokenSource": "manual",
            }
        },
    }
    provider_path = settings / "providers.json"
    provider_path.write_text(json.dumps(providers, separators=(",", ":")) + "\n")
    provider_path.chmod(0o600)
    environment = dict(os.environ)
    environment["CLINE_DATA_DIR"] = str(state)
    environment["CLINE_PROVIDER_SETTINGS_PATH"] = str(provider_path)
    completed = subprocess.run(
        ["/opt/harness/cline-real", *sys.argv[1:]], env=environment, text=True,
        stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
    )
    output = completed.stdout.strip() or f"cline exited {completed.returncode} without output"
    session = hashlib.sha256("\0".join(sys.argv[1:]).encode()).hexdigest()[:24]
    print(json.dumps({
        "schema": "canvasbench.cli-wrapper.v1", "harness": "cline",
        "session_id": f"cline-{session}", "output": output, "exit_code": completed.returncode,
    }, separators=(",", ":")))
    raise SystemExit(completed.returncode)
finally:
    shutil.rmtree(state, ignore_errors=True)
